src/Compilers/CSharp/Portable/Lowering/LocalRewriter/DelegateCacheRewriter.cs

```diff
@@ -0,0 +1,180 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp;
+
+/// <summary>
+/// This type helps rewrite the delegate creations that target static method groups to use a cached instance of delegate.
+/// </summary>
+internal sealed partial class DelegateCreationRewriter
+{
+    private readonly SyntheticBoundNodeFactory _factory;
+    private readonly int _topLevelMethodOrdinal;
+
+    private Dictionary<MethodSymbol, DelegateCacheContainer>? _genericCacheContainers;
+
+    internal DelegateCreationRewriter(SyntheticBoundNodeFactory factory, int topLevelMethodOrdinal)
+    {
+        Debug.Assert(factory.TopLevelMethod is { });
+
+        _factory = factory;
+        _topLevelMethodOrdinal = topLevelMethodOrdinal;
+    }
+
+    internal static bool CanRewrite(MethodSymbol targetMethod, BoundConversion boundConversion)
+        => targetMethod.IsStatic && !boundConversion.IsExtensionMethod;
+
+    internal BoundExpression Rewrite(BoundDelegateCreationExpression boundDelegateCreation, MethodSymbol targetMethod, TypeSymbol delegateType)
```

Comment:
>TypeSymbol delegateType [](http://example.com/codeflow?start=119&length=23)

This looks like a duplicate information that we can easily get from the ```boundDelegateCreation```.

Reply:
Indeed, let me fix it.

Reply:
That was a really nice pick. Fixed, please check 87c448b.